File: NMAP-PREMIUM.py
#!/usr/bin/env python3
# by X2X0/X4X0
import socket
import sys
import threading
import time
import argparse
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime

def print_banner():
    banner = """
    ███╗   ██╗███╗   ███╗ █████╗ ██████╗     ██████╗ ██████╗ ███████╗███╗   ███╗██╗██╗   ██╗███╗   ███╗
    ████╗  ██║████╗ ████║██╔══██╗██╔══██╗    ██╔══██╗██╔══██╗██╔════╝████╗ ████║██║██║   ██║████╗ ████║
    ██╔██╗ ██║██╔████╔██║███████║██████╔╝    ██████╔╝██████╔╝█████╗  ██╔████╔██║██║██║   ██║██╔████╔██║
    ██║╚██╗██║██║╚██╔╝██║██╔══██║██╔═══╝     ██╔═══╝ ██╔══██╗██╔══╝  ██║╚██╔╝██║██║██║   ██║██║╚██╔╝██║
    ██║ ╚████║██║ ╚═╝ ██║██║  ██║██║         ██║     ██║  ██║███████╗██║ ╚═╝ ██║██║╚██████╔╝██║ ╚═╝ ██║
    ╚═╝  ╚═══╝╚═╝     ╚═╝╚═╝  ╚═╝╚═╝         ╚═╝     ╚═╝  ╚═╝╚══════╝╚═╝     ╚═╝╚═╝ ╚═════╝ ╚═╝     ╚═╝
                                                                                        
    """
    print(banner)
    print("    Advanced Port Scanner [Premium Version]")
    print("    Developed in Python  |  https://github.com/X2X0/Nmap-Premium")
    print("    " + "=" * 75)

def get_banner(ip, port, timeout=1):
    try:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.settimeout(timeout)
            s.connect((ip, port))
            s.send(b"GET / HTTP/1.1\r\nHost: localhost\r\n\r\n")
            banner = s.recv(1024).strip()
            return banner.decode('utf-8', errors='ignore') if banner else "No banner"
    except:
        return "No banner"

def scan_port(ip, port, verbose=False, timeout=1):
    try:
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.settimeout(timeout)
            result = s.connect_ex((ip, port))
            if result == 0:
                service = "Unknown"
                try:
                    service = socket.getservbyport(port)
                except:
                    pass
                
                banner = get_banner(ip, port, timeout)
                if verbose:
                    print(f"[+] Port {port}/tcp open - {service}")
                    if banner != "No banner":
                        print(f"    Banner: {banner}")
                return port, True, service, banner
            else:
                if verbose > 1:
                    print(f"[-] Port {port}/tcp closed")
                return port, False, "", ""
    except KeyboardInterrupt:
        print("\n[!] Scan interrupted by user")
        sys.exit(1)
    except socket.gaierror:
        print(f"\n[!] Hostname {ip} could not be resolved")
        sys.exit(1)
    except socket.error:
        print(f"\n[!] Could not connect to server {ip}")
        sys.exit(1)

def port_scan(target, ports, num_threads=100, verbose=False, timeout=1):
    print(f"\n[*] Starting scan on {target}")
    print(f"[*] Scanning {len(ports)} ports with {num_threads} threads")
    print(f"[*] Start time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    
    try:
        ip = socket.gethostbyname(target)
    except socket.gaierror:
        print(f"\n[!] Hostname {target} could not be resolved")
        sys.exit(1)
    
    print(f"[*] Target IP: {ip}\n")
    
    open_ports = []
    start_time = time.time()
    
    with ThreadPoolExecutor(max_workers=num_threads) as executor:
        results = list(executor.map(
            lambda port: scan_port(ip, port, verbose, timeout), 
            ports
        ))
    
    for port, is_open, service, banner in results:
        if is_open:
            open_ports.append((port, service, banner))
    
    elapsed_time = time.time() - start_time
    
    print("\n" + "=" * 60)
    print(f"Scan completed in {elapsed_time:.2f} seconds")
    print(f"Found {len(open_ports)} open ports out of {len(ports)} scanned on {target} ({ip})")
    print("=" * 60)
    
    if open_ports:
        print("\nOpen ports:")
        print("-" * 80)
        print(f"{'PORT':<10}{'SERVICE':<20}{'BANNER'}")
        print("-" * 80)
        for port, service, banner in sorted(open_ports):
            banner_short = banner[:50] + "..." if len(banner) > 50 else banner
            print(f"{port:<10}{service:<20}{banner_short}")
    
    return open_ports

def main():
    print_banner()
    
    parser = argparse.ArgumentParser(description='Advanced Port Scanner in Python')
    parser.add_argument('-t', '--target', required=True, help='Target IP or domain')
    parser.add_argument('-p', '--ports', default='1-1000', help='Ports to scan (e.g.: 22,80,443 or 1-1000)')
    parser.add_argument('-T', '--threads', type=int, default=100, help='Number of threads (default: 100)')
    parser.add_argument('-v', '--verbose', action='count', default=0, help='Verbosity level')
    parser.add_argument('-o', '--output', help='Output file to save results')
    parser.add_argument('--timeout', type=float, default=1, help='Timeout in seconds (default: 1)')
    
    args = parser.parse_args()
    
    ports_to_scan = []
    for port_range in args.ports.split(','):
        if '-' in port_range:
            start, end = map(int, port_range.split('-'))
            ports_to_scan.extend(range(start, end + 1))
        else:
            ports_to_scan.append(int(port_range))
    
    ports_to_scan = sorted(list(set(ports_to_scan)))
    
    open_ports = port_scan(
        args.target, 
        ports_to_scan, 
        num_threads=args.threads, 
        verbose=args.verbose,
        timeout=args.timeout
    )
    
    if args.output and open_ports:
        try:
            with open(args.output, 'w') as f:
                f.write(f"Scan results for {args.target} ({socket.gethostbyname(args.target)})\n")
                f.write(f"Date: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
                f.write(f"Ports scanned: {args.ports}\n\n")
                f.write(f"{'PORT':<10}{'SERVICE':<20}{'BANNER'}\n")
                f.write("-" * 80 + "\n")
                for port, service, banner in sorted(open_ports):
                    f.write(f"{port:<10}{service:<20}{banner}\n")
            print(f"\n[+] Results saved to {args.output}")
        except Exception as e:
            print(f"\n[!] Error saving results: {e}")

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n[!] Scan interrupted by user")
        sys.exit(1)

File: NMAP-PREMIUM.java
import java.io.*;
import java.net.*;
import java.util.*;
import java.util.concurrent.*;
import java.text.SimpleDateFormat;

public class PortScanner {

    public static void printBanner() {
        String banner = """
        ███╗   ██╗███╗   ███╗ █████╗ ██████╗     ██████╗ ██████╗ ███████╗███╗   ███╗██╗██╗   ██╗███╗   ███╗
        ████╗  ██║████╗ ████║██╔══██╗██╔══██╗    ██╔══██╗██╔══██╗██╔════╝████╗ ████║██║██║   ██║████╗ ████║
        ██╔██╗ ██║██╔████╔██║███████║██████╔╝    ██████╔╝██████╔╝█████╗  ██╔████╔██║██║██║   ██║██╔████╔██║
        ██║╚██╗██║██║╚██╔╝██║██╔══██║██╔═══╝     ██╔═══╝ ██╔══██╗██╔══╝  ██║╚██╔╝██║██║██║   ██║██║╚██╔╝██║
        ██║ ╚████║██║ ╚═╝ ██║██║  ██║██║         ██║     ██║  ██║███████╗██║ ╚═╝ ██║██║╚██████╔╝██║ ╚═╝ ██║
        ╚═╝  ╚═══╝╚═╝     ╚═╝╚═╝  ╚═╝╚═╝         ╚═╝     ╚═╝  ╚═╝╚══════╝╚═╝     ╚═╝╚═╝ ╚═════╝ ╚═╝     ╚═╝
        """;
        System.out.println(banner);
        System.out.println("    Advanced Port Scanner [Premium Version]");
        System.out.println("    Developed in Java  |  https://github.com/X2X0/Nmap-Premium");
        System.out.println("    " + "=".repeat(75));
    }

    public static String getBanner(String ip, int port, int timeout) {
        try (Socket socket = new Socket()) {
            socket.connect(new InetSocketAddress(ip, port), timeout);
            socket.getOutputStream().write("GET / HTTP/1.1\r\nHost: localhost\r\n\r\n".getBytes());
            byte[] banner = new byte[1024];
            int length = socket.getInputStream().read(banner);
            return new String(banner, 0, length).trim();
        } catch (IOException e) {
            return "No banner";
        }
    }

    public static class ScanResult {
        int port;
        boolean isOpen;
        String service;
        String banner;

        public ScanResult(int port, boolean isOpen, String service, String banner) {
            this.port = port;
            this.isOpen = isOpen;
            this.service = service;
            this.banner = banner;
        }
    }

    public static ScanResult scanPort(String ip, int port, boolean verbose, int timeout) {
        try (Socket socket = new Socket()) {
            socket.connect(new InetSocketAddress(ip, port), timeout);
            String service = "Unknown";
            try {
                service = InetAddress.getByName("localhost").getCanonicalHostName();
            } catch (UnknownHostException ignored) {}

            String banner = getBanner(ip, port, timeout);
            if (verbose) {
                System.out.println("[+] Port " + port + "/tcp open - " + service);
                if (!"No banner".equals(banner)) {
                    System.out.println("    Banner: " + banner);
                }
            }
            return new ScanResult(port, true, service, banner);
        } catch (IOException e) {
            if (verbose) {
                System.out.println("[-] Port " + port + "/tcp closed");
            }
            return new ScanResult(port, false, "", "");
        }
    }

    public static List<ScanResult> portScan(String target, List<Integer> ports, int numThreads, boolean verbose, int timeout) {
        System.out.println("\n[*] Starting scan on " + target);
        System.out.println("[*] Scanning " + ports.size() + " ports with " + numThreads + " threads");
        System.out.println("[*] Start time: " + new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date()));

        String ip;
        try {
            ip = InetAddress.getByName(target).getHostAddress();
        } catch (UnknownHostException e) {
            System.out.println("\n[!] Hostname " + target + " could not be resolved");
            return new ArrayList<>();
        }

        System.out.println("[*] Target IP: " + ip + "\n");

        List<ScanResult> openPorts = new ArrayList<>();
        long startTime = System.nanoTime();

        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        List<Future<ScanResult>> futures = new ArrayList<>();
        for (int port : ports) {
            futures.add(executor.submit(() -> scanPort(ip, port, verbose, timeout)));
        }

        for (Future<ScanResult> future : futures) {
            try {
                ScanResult result = future.get();
                if (result.isOpen) {
                    openPorts.add(result);
                }
            } catch (InterruptedException | ExecutionException e) {
                e.printStackTrace();
            }
        }
        executor.shutdown();

        long elapsedTime = (System.nanoTime() - startTime) / 1_000_000;
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Scan completed in " + elapsedTime + " ms");
        System.out.println("Found " + openPorts.size() + " open ports out of " + ports.size() + " scanned on " + target);
        System.out.println("=" .repeat(60));

        if (!openPorts.isEmpty()) {
            System.out.println("\nOpen ports:");
            System.out.println("-".repeat(80));
            System.out.printf("%-10s%-20s%-50s%n", "PORT", "SERVICE", "BANNER");
            System.out.println("-".repeat(80));
            for (ScanResult result : openPorts) {
                String bannerShort = result.banner.length() > 50 ? result.banner.substring(0, 50) + "..." : result.banner;
                System.out.printf("%-10d%-20s%-50s%n", result.port, result.service, bannerShort);
            }
        }

        return openPorts;
    }

    public static void main(String[] args) {
        printBanner();

        String target = args[0];
        String portRange = args[1];
        int threads = Integer.parseInt(args[2]);
        boolean verbose = args.length > 3 && args[3].equals("-v");
        int timeout = 1000;

        List<Integer> portsToScan = new ArrayList<>();
        for (String port : portRange.split(",")) {
            if (port.contains("-")) {
                String[] range = port.split("-");
                int start = Integer.parseInt(range[0]);
                int end = Integer.parseInt(range[1]);
                for (int i = start; i <= end; i++) {
                    portsToScan.add(i);
                }
            } else {
                portsToScan.add(Integer.parseInt(port));
            }
        }

        portsToScan = new ArrayList<>(new HashSet<>(portsToScan));
        Collections.sort(portsToScan);

        List<ScanResult> openPorts = portScan(target, portsToScan, threads, verbose, timeout);

        if (args.length > 4) {
            String outputFile = args[4];
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile))) {
                writer.write("Scan results for " + target + " (" + InetAddress.getByName(target).getHostAddress() + ")\n");
                writer.write("Date: " + new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date()) + "\n");
                writer.write("Ports scanned: " + portRange + "\n\n");
                writer.write(String.format("%-10s%-20s%-50s%n", "PORT", "SERVICE", "BANNER"));
                writer.write("-".repeat(80) + "\n");
                for (ScanResult result : openPorts) {
                    writer.write(String.format("%-10d%-20s%-50s%n", result.port, result.service, result.banner));
                }
                System.out.println("\n[+] Results saved to " + outputFile);
            } catch (IOException e) {
                System.out.println("\n[!] Error saving results: " + e.getMessage());
            }
        }
    }
                             }
